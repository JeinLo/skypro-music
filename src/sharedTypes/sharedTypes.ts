export interface TrackType {
  _id: number;
  name: string;
  author: string;
  release_date: string;
  genre: string[];
  duration_in_seconds: number;
  album: string;
  logo: null | string;
  track_file: string;
  starred_user: number[];
  isLiked?: boolean;
}

export interface PlaylistSelectionType {
  _id: number;
  name: string;
  items: number[];
  owner: number[];
  __v: number;
}