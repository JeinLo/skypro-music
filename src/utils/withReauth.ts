import { AxiosError } from 'axios';
import { refreshToken } from '@/services/auth/tokenApi';
import { AppDispatch } from '@/store/store';
import { setAccessToken } from '@/store/features/authSlice';

export const withReauth = async <T>(
  apiFunction: (access: string) => Promise<T>,
  refresh: string,
  dispatch: AppDispatch,
): Promise<T> => {
  try {
    return await apiFunction('');
  } catch (error) {
    const axiosError = error as AxiosError;
    if (axiosError.response?.status === 401) {
      try {
        const newTokens = await refreshToken(refresh);
        dispatch(setAccessToken(newTokens.access));
        return await apiFunction(newTokens.access);
      } catch (refreshError) {
        throw refreshError;
      }
    }
    throw error;
  }
};