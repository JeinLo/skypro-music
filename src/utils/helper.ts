import { TrackType } from "@/sharedTypes/sharedTypes";

export function getUniqueValuesByKey(
  arr: TrackType[],
  key: keyof TrackType,
): string[] {
  const uniqueValues = new Set<string>();

  arr.forEach((item) => {
    const value = item[key];

    if (Array.isArray(value)) {
      value.forEach((v) => {
        if (v) {
          uniqueValues.add(String(v)); // ← Добавлено String() для string[]
        }
      });
    } else if (typeof value === 'string') {
      uniqueValues.add(value);
    } else if (typeof value === 'number') {
      uniqueValues.add(String(value)); // ← Добавлено для number (_id, duration_in_seconds)
    }
  });

  return Array.from(uniqueValues);
}

export function formatTime(time: number) {
  const minutes = Math.floor(time / 60);
  const inputSeconds = Math.floor(time % 60);
  const outputSeconds = inputSeconds < 10 ? `0${inputSeconds}` : inputSeconds;

  return `${minutes}:${outputSeconds}`;
}

export const getTimePanel = (currentTime: number, totalTime: number | undefined) => {
  return `${formatTime(currentTime)} / ${totalTime ? formatTime(totalTime) : '--:--'}`;
};