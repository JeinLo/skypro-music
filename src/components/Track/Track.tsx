'use client';

import Link from 'next/link';
import classNames from 'classnames';
import styles from './Track.module.css';
import { TrackType } from '@/sharedTypes/sharedTypes';
import { formatTime } from '@/utils/helper';
import { useAppDispatch, useAppSelector } from '@/store/store';
import { setCurrentTrack, toggleFavorite } from '@/store/features/trackSlice';
import { useState, useEffect } from 'react';

type TrackProps = {
  track: TrackType;
  playlist: TrackType[];
};

export default function Track({ track, playlist }: TrackProps) {
  const dispatch = useAppDispatch();
  const isPlay = useAppSelector((state) => state.tracks.isPlay);
  const currentTrack = useAppSelector((state) => state.tracks.currentTrack);
  const favoriteTrackIds = useAppSelector((state) => state.tracks.favoriteTrackIds);
  const [isFavorite, setIsFavorite] = useState(favoriteTrackIds.includes(track._id));

  useEffect(() => {
    // Синхронизация favoriteTrackIds с localStorage
    const savedFavorites = localStorage.getItem('favoriteTrackIds');
    if (savedFavorites) {
      const favoriteIds = JSON.parse(savedFavorites) as number[];
      favoriteIds.forEach((id) => {
        if (!favoriteTrackIds.includes(id)) {
          dispatch(toggleFavorite(id));
        }
      });
    }
  }, [dispatch]);

  useEffect(() => {
    // Сохранение favoriteTrackIds в localStorage
    localStorage.setItem('favoriteTrackIds', JSON.stringify(favoriteTrackIds));
    setIsFavorite(favoriteTrackIds.includes(track._id));
  }, [favoriteTrackIds, track._id]);

  const isCurrentTrack = currentTrack?._id === track._id;

  const onClickCurrentTrack = () => {
    dispatch(setCurrentTrack({ track, playlist }));
  };

  const toggleFavoriteTrack = () => {
    dispatch(toggleFavorite(track._id));
  };

  return (
    <div className={styles.playlist__item} onClick={onClickCurrentTrack}>
      <div className={styles.playlist__track}>
        <div className={styles.track__title}>
          <div className={styles.track__titleImage}>
            <svg className={styles.track__titleSvg}>
              <use xlinkHref="/img/icon/sprite.svg#icon-note"></use>
            </svg>
            {isCurrentTrack && (
              <span
                className={classNames(styles.currentTrackDot, {
                  [styles.pulsating]: isPlay,
                })}
              />
            )}
          </div>
          <div className={styles.track__titleText}>
            <Link
              className={classNames(styles.track__titleLink, {
                [styles.activeTrack]: isCurrentTrack,
              })}
              href=""
            >
              {track.name} <span className={styles.track__titleSpan}></span>
            </Link>
          </div>
        </div>
        <div className={styles.track__author}>
          <Link className={styles.track__authorLink} href="">
            {track.author}
          </Link>
        </div>
        <div className={styles.track__album}>
          <Link className={styles.track__albumLink} href="">
            {track.album}
          </Link>
        </div>
        <div className={styles.track__time}>
          <svg
            className={classNames(styles.track__timeSvg, { [styles.active]: isFavorite })}
            onClick={(e) => {
              e.stopPropagation();
              toggleFavoriteTrack();
            }}
          >
            <use xlinkHref={`/img/icon/sprite.svg#icon-${isFavorite ? 'like-filled' : 'like'}`} />
          </svg>
          <span className={styles.track__timeText}>
            {formatTime(track.duration_in_seconds)}
          </span>
        </div>
      </div>
    </div>
  );
}