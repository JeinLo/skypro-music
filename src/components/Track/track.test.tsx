import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { data } from '@/data';
import { TrackType } from '@/sharedTypes/sharedTypes';
import ReduxProvider from '@/store/ReduxProvider';
import Track from './Track';
import { formatTime } from '@/utils/helper';

const mockTracks: TrackType[] = data;
const mockTrack: TrackType = data[0];

describe('Track component', () => {
  test('Отрисовка данных трека', () => {
   render(<ReduxProvider>
   <Track track={mockTrack} playlist={mockTracks} />
      </ReduxProvider>
   )
    expect (screen.getAllByText(mockTrack.author).length).toBeGreaterThan(0);
    expect (screen.getAllByText(mockTrack.name).length).toBeGreaterThan(0);
    expect (screen.getAllByText(formatTime(mockTrack.duration_in_seconds)).length).toBeGreaterThan(0);
  });
  test('Снапшот трека', () => {
  const { container } = render(
    <ReduxProvider>
      <Track track={mockTrack} playlist={mockTracks} />
    </ReduxProvider>
  );
  expect(container).toMatchSnapshot();
});
});