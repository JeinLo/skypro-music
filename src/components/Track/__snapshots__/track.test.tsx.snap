// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Track component Снапшот трека 1`] = `
<div>
  <div
    class="playlist__item"
  >
    <div
      class="playlist__track"
    >
      <div
        class="track__title"
      >
        <div
          class="track__titleImage"
        >
          <svg
            class="track__titleSvg"
          >
            <use
              xlink:href="/img/icon/sprite.svg#icon-note"
            />
          </svg>
        </div>
        <div
          class="track__titleText"
        >
          <a
            class="track__titleLink"
            href=""
          >
            Chase
             
            <span
              class="track__titleSpan"
            />
          </a>
        </div>
      </div>
      <div
        class="track__author"
      >
        <a
          class="track__authorLink"
          href=""
        >
          Alexander Nakarada
        </a>
      </div>
      <div
        class="track__album"
      >
        <a
          class="track__albumLink"
          href=""
        >
          Chase
        </a>
      </div>
      <div
        class="track__time"
      >
        <svg
          class="track__timeSvg"
        >
          <use
            xlink:href="/img/icon/sprite.svg#icon-like"
          />
        </svg>
        <span
          class="track__timeText"
        >
          3:25
        </span>
      </div>
    </div>
  </div>
</div>
`;
